design-proposals/vm-backend-storage.md

```diff
@@ -0,0 +1,73 @@
+# Overview
+Today, anything that happens on the host side of VMIs (i.e. virt-launcher) is transient.  
+Any change a VMI makes outside of its own file-system is lost after a shutdown.  
+Such changes include EFI firmware settings and information stored inside the TPM.  
+This design provides a way to persist small VM-specific files used by virt-launcher.
+
+## Motivation
+Though we now support exposing an emulated TPM device to guests (vTPM), using the device in any meaningful way would require persisting data across reboots.  
+As we do not currently have a way to do that, all data stored by the VMI into the vTPM is lost after a shutdown (a soft reboot keeps the same virt-launcher, avoiding losses).
+
+## Goals
+- Being able to persist modifications of various emulated devices across reboots for a given VM
+- Enable live migration of VMs that use backend storage
+- Avoid consuming large amounts of storage
+- Prevent virt-launcher for VM X from seeing data that belongs to VM Y
+
+## Non Goals
+- Persisting everything. Some information that would usually persist on a non-ephemeral host just don't matter to us.
+- Supporting user modifications of the backend storage. Users whishing to inject their own files should do so at their own risk.
+
+## Definition of Users
+This feature is intended for end-users who wish to preserve the state of some emulated device across reboots.
+
+## User Stories
+As a VM user, I want to be able to store the encryption key of my VMs partitions inside the vTPM (example: BitLocker)
+
+## Repos
+kubevirt/kubevirt
+
+# Design
+The storage used for this would be 1 PVC per VM, and would be created on first start of a VM that needs backend storage.  
+This PVC would be RWO, about 2 MiB in size (still TBD), and use [csi-hostpath](https://github.com/kubevirt/hostpath-provisioner).  
+The PVC would be added to the virt-launcher compute container as a volume, but never assigned to the libvirt domain (the guest VM).  
+The PVC would be mounted in a directory at the root of the filesystem. Appropriate symbolic links will be created before starting the libvirt domain.
```

Comment:
Can you add an example of such symlinks and maybe explain what is the motivation for it?

Reply:
I actually don't like the symlinks idea, but that's the only way I could think of to make the files appear to be in the right place.
When there's a libvirt domain option to override the location we can just point to the actual file, but for example TPM doesn't seem to have such an option...
If you have other ideas I'm all for it, for example I've seen overlay2-mounted files, not directories, inside containers. I don't understand how that works, since you're only supposed to be able to mount directories, but maybe that mechanism could be used to our advantage?...

Reply:
@xpivarc did you get a chance to test your idea? Is that something I could help with? Thanks!

Reply:
My idea was to use multiple moutPaths for one volume. Would that help?

Reply:
Yeah, but would that require RWX? I guess it would have the undesirable side effect of showing all files in all mount points, but I still prefer that over symlinks.
We can also simplify all this by having 1 volume per feature/mountpoint and be done with it (e.g. 1 for TPM, 1 of OVMF, ....). That's probably our best bet when I think about it.

Reply:
I think the RWX is a must for another reason https://github.com/kubevirt/community/pull/170#discussion_r848081613. 

The idea would not require RWX afaik. You could use subPaths and limit the view. I think the best would be to try it out. Unfortunately, I didn't try it yet.